ast/src/analyzed/mod.rs

```diff
@@ -565,6 +565,65 @@ pub fn type_from_definition(
     }
 }
 
+/// Data structure to help with finding the correct implementation of a trait function
+/// given a list of type arguments.
+#[derive(Default, Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct SolvedTraitImpls {
+    impls: HashMap<String, HashMap<Vec<Type>, ImplData>>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+struct ImplData {
+    index: usize,
+    function: Arc<Expression>,
+}
+
+impl SolvedTraitImpls {
+    /// Returns an index into the list of trait implementations such that the corresponding
+    /// trait implementation contains the matching function for a given trait function name
+    /// and type arguments. It returns an index into the list provided by `Analyzed`.
+    pub fn resolve_trait_impl_index(&self, trait_function_name: &str, type_args: &[Type]) -> usize {
+        self.impls[trait_function_name][type_args].index
+    }
+
+    /// Returns the correct trait impl function for a given trait function name and type arguments,
+    /// if it is stored here, otherwise returns None.
+    pub fn try_resolve_trait_function(
+        &self,
+        trait_function_name: &str,
+        type_args: &[Type],
+    ) -> Option<&Expression> {
+        self.impls
+            .get(trait_function_name)
+            .and_then(|map| map.get(type_args))
+            .map(|impl_data| impl_data.function.as_ref())
+    }
+
+    /// Returns the correct trait impl function for a given trait function name and type arguments.
+    /// It returns just the function of the trait impl.
+    pub fn resolve_trait_function(
+        &self,
+        trait_function_name: &str,
+        type_args: &[Type],
+    ) -> &Expression {
+        self.try_resolve_trait_function(trait_function_name, type_args)
+            .unwrap()
+    }
+
+    pub fn insert(
+        &mut self,
+        trait_function_name: String,
+        type_args: Vec<Type>,
+        index: usize,
+        function: Arc<Expression>,
+    ) {
+        self.impls
+            .entry(trait_function_name)
+            .or_default()
+            .insert(type_args, ImplData { index, function });
```

Comment:
We never merge the code to check for overlaps. Couldn't we have a problem if this `.insert(type_args, ...)` overlaps?

Reply:
Yes, but that's why we still need to do the overlap check :)